functorch/test/test_eager_transforms.py

```diff
@@ -2014,6 +2014,146 @@ def push_jvp(dummy, x):
+        x = orig_x.clone()
+        xt = orig_xt.clone()
+        y = torch.randn(2, B, 3, device=device)
+        yt = torch.randn(2, B, 3, device=device)
+        out, out_tangent = vmap(push_jvp(f), in_dims=(2, 2, 1, 1))(x, xt, y, yt)
+
+        self.assertEqual(out, y.movedim(1, 0))
+        self.assertEqual(out_tangent, yt.movedim(1, 0))
+
+        expected = orig_x.movedim(2, 0).clone()
+        expected[:, :, ::2] = y.movedim(1, 0)
+        self.assertEqual(x.movedim(2, 0), expected)
+
+    # Case 3 in [Forward Grad View/inplace]
+    def test_all_dual_base_inplace(self, device):
+        B = 2
+
+        def push_jvp(f):
+            def inner(x, xt, y, yt):
+                return jvp(f, (x, y), (xt, yt))
+            return inner
+
+        # Case 3: with view, propagate from base to view
+        def f(x, y):
+            view = x[0, ::2]
+            x.copy_(y)
+            return x
+
+        x = torch.randn(2, B, 6, device=device)
+        xt = torch.randn(2, 6, B, device=device)
+        y = torch.randn(2, B, 6, device=device)
+        yt = torch.randn(2, B, 6, device=device)
+        out, out_tangent = vmap(push_jvp(f), in_dims=(1, 2, 1, 1))(x, xt, y, yt)
+
+        self.assertEqual(out, y.movedim(1, 0))
+        self.assertEqual(out_tangent, yt.movedim(1, 0))
+
+    # Case 4 in [Forward Grad View/inplace]
+    def test_right_dual_view_prop(self, device):
+        B = 2
+
+        # Changes on the view must propagate to its base. Also:
+        # - x is a regular Tensor
+        # - y is a dual tensor
+        def f(x, y):
+            x = x.clone()
+            view = x[0]
+            view.copy_(y)
+            return view, x
+
+        def push_jvp(x, y, yt):
+            return jvp(partial(f, x), (y,), (yt,))
+
+        x = torch.randn(2, B, 6, device=device)
+        y = torch.randn(6, B, device=device)
+        yt = torch.randn(6, B, device=device)
+        outs, tangents = vmap(push_jvp, in_dims=(1, 1, 1))(x, y, yt)
+
+        self.assertEqual(outs[0], y.movedim(1, 0))
+        self.assertEqual(tangents[0], yt.movedim(1, 0))
+
+        expected_tangent_1 = torch.zeros_like(x).movedim(1, 0)
+        expected_tangent_1[:, 0].copy_(yt.movedim(1, 0))
+        self.assertEqual(tangents[1], expected_tangent_1)
+
+    # Case 5 in [Forward Grad View/inplace]
+    def test_right_dual_base_prop(self, device):
+        B = 2
+
+        # Changes on the base must propagate on all its views. Also:
+        # - x is a regular Tensor
+        # - y is a dual tensor
+        def f(x, y):
+            x = x.clone()
+            view = x[0]
+            x.copy_(y)
+            return view, x
+
+        def push_jvp(x, y, yt):
+            return jvp(partial(f, x), (y,), (yt,))
+
+        x = torch.randn(2, B, 6)
+        y = torch.randn(2, 6, B)
+        yt = torch.randn(2, 6, B)
+        outs, tangents = vmap(push_jvp, in_dims=(1, 2, 2))(x, y, yt)
+
+        self.assertEqual(outs[1], y.movedim(2, 0))
```

Comment:
same question as above but inverted for outs[0]

Reply:
(y)